execution/executor/src/db_bootstrapper.rs

```diff
@@ -30,27 +34,47 @@ pub fn bootstrap_db<V: VMExecutor>(
     tree_state: TreeState,
     genesis_txn: &Transaction,
 ) -> Result<Waypoint> {
+    let new_genesis_version = tree_state.num_transactions;
     let mut executor = Executor::<V>::new_on_unbootstrapped_db(db.clone(), tree_state);
 
-    let block_id = HashValue::zero(); // match with the id in BlockInfo::genesis(...)
+    let block_id = HashValue::zero();
+    let epoch = if new_genesis_version == 0 {
+        GENESIS_EPOCH
+    } else {
+        db.reader.get_latest_ledger_info()?.ledger_info().epoch() + 1
+    };
 
     // Create a block with genesis_txn being the only txn. Execute it then commit it immediately.
     let result =
         executor.execute_block((block_id, vec![genesis_txn.clone()]), *PRE_GENESIS_BLOCK_ID)?;
 
     let root_hash = result.root_hash();
-    let validator_set = result
+    let next_validator_set = result
         .validators()
-        .clone()
-        .expect("Genesis transaction must emit a validator set.");
+        .as_ref()
+        .ok_or_else(|| format_err!("Genesis transaction must emit a validator set."))?;
 
-    let ledger_info_with_sigs = LedgerInfoWithSignatures::genesis(root_hash, validator_set.clone());
+    let ledger_info_with_sigs = LedgerInfoWithSignatures::new(
+        LedgerInfo::new(
+            BlockInfo::new(
+                epoch,
+                GENESIS_ROUND,
+                block_id,
+                root_hash,
+                new_genesis_version,
+                GENESIS_TIMESTAMP_USECS,
```

Comment:
it's not related to this PR but this is broken, can you fix this by reading LibraTimestamp?